src/v/storage/compaction_reducers.cc

```diff
@@ -149,6 +149,25 @@ copy_data_segment_reducer::filter(model::record_batch batch) {
             batch, r, batch.record_count() == records_seen, offset_deltas);
       });
 
+    if (batch.last_offset() == _segment_last_offset && offset_deltas.empty()) {
+        // last batch in the segment has been compacted away.
+        // This is most likely caused by aborted data batches getting compacted
+        // away during self compaction of the segment if they are the last batch
+        // in the segment. We install a placeholder batch of same size to retain
+        // contiguousness of the offset space.
+        record_batch_builder builder(
+          model::record_batch_type::compaction_placeholder,
+          batch.base_offset());
+        for (auto i = batch.base_offset(); i <= batch.last_offset(); i++) {
+            builder.add_raw_kv({}, {});
+        }
```

Comment:
Might not be worth optimizing, but I don't think this needs to actually have the right number of records, as long as the offset boundaries of the batch are okay.

Reply:
part of the reason I didn't do this originally is because the record batch API seems to convoluted for a use case like this (copy the header, reset crc etc).. anyway I switched it now.